### PR TITLE
[WFCORE-6739] Allow ordering of ServerActivity execution by registerin…

### PR DESCRIPTION
**Pull request by bstansberry**
…g them in distinct, ordered, 'execution groups'

https://issues.redhat.com/browse/WFCORE-6739

This is a draft pending further discussion, particularly of how suitable this is for the transaction subsystem's implementation of https://issues.redhat.com/browse/WFLY-17742.

@jmfinelli FYI.

### CODE REVIEW

File: server/src/main/java/org/jboss/as/server/suspend/SuspendController.java [hunk 1]
```diff
@@ -75,20 +81,32 @@ public synchronized void suspend(long timeoutMillis) {
         for(OperationListener listener: new ArrayList<>(operationListeners)) {
             listener.suspendStarted();
         }
-        outstandingCount = activities.size();
-        if (outstandingCount == 0) {
+        groupsCount = activities.size();
```
Comment (jmfinelli) [inline server/src/main/java/org/jboss/as/server/suspend/SuspendController.java:84]: Shouldn't this be
```suggestion
        // Work out the number of ServerActivity implementations 
        activities.forEach((ranking,executionGroup) -> groupsCount =+ executionGroup.size());
```

Or, we could also add a `groupsCount++` in `registerActivity(ServerActivity)`

**Review by jmfinelli — COMMENTED** [review]
I'm not sure why this is a DRAFT PR :-) It looks super good to me.
I just left a comment in the code.

Thank you very much @bstansberry for taking care of this

File: server/src/main/java/org/jboss/as/server/suspend/SuspendController.java (hunk 1 above)
Comment (bstansberry) [reply server/src/main/java/org/jboss/as/server/suspend/SuspendController.java:84]: I don't think so. This var is meant to track the number of groups, not the number of activities. When each *group* completes preSuspend work it signals preSuspendGroupCallBack. When preSuspendGroupCallBack get groupsCount signals, it proceeds with the 'suspend' phase.

I should probably rename the 'activities' field to 'activitiesByGroup'.

**Comment by bstansberry** [comment]
Thanks for the review, @jmfinelli. 

It's draft for a couple reasons. :-)

The biggest is that we're doing this to meet your use case, so before having other folks waste time looking at it, I wanted to make sure you thought it would. When we talked I thought we understood each other and something like this should solve your problem, but maybe I misunderstood or you thought more and don't think it will help etc. Sounds like it's ok though, so that's good.

The other is that I should write a unit test. This works fine it that it doesn't break existing tests but that doesn't prove it actually works. :)

Looking at SuspendController it seems very unit-testable, and the basic behavior that needs testing is straightforward (activities are invoked in the expected order regardless of the order of calls to registerActivity).

File: server/src/main/java/org/jboss/as/server/suspend/SuspendController.java (hunk 1 above)
Comment (jmfinelli) [reply server/src/main/java/org/jboss/as/server/suspend/SuspendController.java:84]: Yep, I see it now. Sorry for the noise.

**Comment by jmfinelli** [comment]
Yep, it look super good to me and it reflects exactly what we talked about during the meeting :-) I think this is the last piece of the puzzle :partying_face: I also declared a dependency in the [proposal](https://github.com/wildfly/wildfly-proposals/pull/520/files#diff-e4492f568858140c27f70bb443a285510631fb5836e49b4a8684b89dcd2cec2fR104).

**Comment by bstansberry** [comment]
@jmfinelli I added the test and moved this out of draft state.

**Comment by bstansberry** [comment]
Thanks @jmfinelli

**Pull request merged**